### PR TITLE
Implements `make_symbol` system macro

### PR DESCRIPTION
**Pull request by zslayton**
By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: src/lazy/encoder/value_writer.rs [hunk 1]
```diff
@@ -36,11 +36,9 @@ pub(crate) mod internal {
...
 }
 
-/// The context in which a given value is found: either a container or the top level.
-//
-// This type is a public-facing version of `MakeValueWriter`.
-// It allows applications to refer to the value writer's associated types without exposing
-// the `make_value_writer` method.
+/// A writer which can encode nested values.
+///
+/// Implementors include top-level writers, container writers, and e-expression writers.
```
Comment (zslayton) [inline src/lazy/encoder/value_writer.rs:41]: 🗺️ I had this doc comment in my workspace as part of #867 but didn't push it. 🤦 Slipping it into this PR.

File: src/lazy/expanded/macro_evaluator.rs [hunk 2]
```diff
@@ -1041,8 +1027,48 @@ impl<'top, D: Decoder> MakeStringExpansion<'top, D> {
...
+        }
+    }
+
+    /// Yields the next [`ValueExpr`] in this `make_string` macro's evaluation.
+    pub fn next(
+        &mut self,
+        context: EncodingContextRef<'top>,
+        _environment: Environment<'top, D>,
+    ) -> IonResult<MacroExpansionStep<'top, D>> {
+        let constructed_text = self.arguments.concatenate_text_arguments(context)?;
```
Comment (zslayton) [inline src/lazy/expanded/macro_evaluator.rs:1071]: 🗺️ `make_string` now just calls `MakeTextExpansion::concatenate_text_arguments` and then massages the resulting `&'bump str` into a `ValueRef`.

File: src/lazy/expanded/macro_evaluator.rs [hunk 3]
```diff
@@ -1058,6 +1084,58 @@ impl<'top, D: Decoder> MakeStringExpansion<'top, D> {
...
+        }
+    }
+
+    /// Yields the next [`ValueExpr`] in this `make_symbol` macro's evaluation.
+    pub fn next(
+        &mut self,
+        context: EncodingContextRef<'top>,
+        _environment: Environment<'top, D>,
+    ) -> IonResult<MacroExpansionStep<'top, D>> {
+        let constructed_text = self.arguments.concatenate_text_arguments(context)?;
```
Comment (zslayton) [inline src/lazy/expanded/macro_evaluator.rs:1123]: 🗺️ The new `make_symbol` impl calls `MakeTextExpansion::concatenate_text_arguments` and then massages the resulting `&'bump str` into a `ValueRef`.

**Review by zslayton — COMMENTED** [review]
🗺️ PR Tour 🧭

File: src/lazy/expanded/macro_evaluator.rs [hunk 4]
```diff
@@ -979,40 +982,23 @@ impl<'top, D: Decoder> ExprGroupExpansion<'top, D> {
...
-///   (:make_string foo bar)                  => "foobar"
-///   (:make_string "foo" bar)                => "foobar"
-///   (:make_string "first_" $4)              => "first_name"
-///   (:make_string (:values "first" "_") $4) => "first_name"
-///   (:make_string)                          => ""
-///   (:make_string "foo" 7)                  => Error
+/// Evaluation logic shared by the `make_string` and `make_symbol` macros.
 #[derive(Copy, Clone, Debug)]
-pub struct MakeStringExpansion<'top, D: Decoder> {
+pub struct MakeTextExpansion<'top, D: Decoder> {
```
[the remark this answers is not in the document]
Comment (popematt) [reply src/lazy/expanded/macro_evaluator.rs:987]: Do you need to have separate implementations that extend/wrap this? The only difference is at the end where it constructs either a Symbol or a String. Could you just have two different constructors that set it up to use either 
`|constructed_text| ValueRef::String(StrRef::from(constructed_text))` or `|constructed_text| ValueRef::Symbol(SymbolRef::from(constructed_text))` depending on which one you need?

[the remark this answers is not in the document]
Comment (zslayton) [reply src/lazy/expanded/macro_evaluator.rs:987]: Yeah, I could do that. I was trying to keep the doc comments for `make_string`, but I guess that's not worth too much.

**Review by popematt — APPROVED** [review]

**Pull request merged**